src/main/java/org/zalando/nakadi/service/CursorOperationsService.java

```diff
@@ -0,0 +1,169 @@
+import org.zalando.nakadi.domain.NakadiCursor;
+import org.zalando.nakadi.domain.NakadiCursorDistanceQuery;
+import org.zalando.nakadi.domain.NakadiCursorDistanceResult;
+import org.zalando.nakadi.domain.ShiftedNakadiCursor;
+import org.zalando.nakadi.domain.Timeline;
+import org.zalando.nakadi.exceptions.NakadiException;
+import org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation;
+import org.zalando.nakadi.service.timeline.TimelineService;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.CURSORS_WITH_DIFFERENT_PARTITION;
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.INVERTED_OFFSET_ORDER;
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.INVERTED_TIMELINE_ORDER;
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.PARTITION_NOT_FOUND;
+import static org.zalando.nakadi.exceptions.runtime.InvalidCursorOperation.Reason.TIMELINE_NOT_FOUND;
+
+public class CursorOperationsService {
+    public static final int CURSOR_OFFSET_LENGTH = 18;
+
+    private final TimelineService timelineService;
+
+    public CursorOperationsService(final TimelineService timelineService) {
+        this.timelineService = timelineService;
+    }
+
+    public List<NakadiCursorDistanceResult> calculateDistance(final List<NakadiCursorDistanceQuery> cursors)
+        throws InvalidCursorOperation {
+        return cursors.stream().map(this::calculateDistance).collect(Collectors.toList());
+    }
+
+    public NakadiCursorDistanceResult calculateDistance(final NakadiCursorDistanceQuery query) {
+        final NakadiCursor initialCursor = query.getInitialCursor();
+        final NakadiCursor finalCursor = query.getFinalCursor();
+
+        if (!initialCursor.getPartition().equals(finalCursor.getPartition())) {
+            throw new InvalidCursorOperation(CURSORS_WITH_DIFFERENT_PARTITION);
+        }
+
+        if (initialCursor.getTimeline().getOrder().equals(finalCursor.getTimeline().getOrder())) {
+            final long distance = getDistanceSameTimeline(query, initialCursor.getOffset(), finalCursor.getOffset());
+            return new NakadiCursorDistanceResult(query, distance);
+        } else if (initialCursor.getTimeline().getOrder() > finalCursor.getTimeline().getOrder()) {
+            throw new InvalidCursorOperation(INVERTED_TIMELINE_ORDER);
+        } else {
+            final long distance = getDistanceDifferentTimelines(query, initialCursor, finalCursor);
+            return new NakadiCursorDistanceResult(query, distance);
+        }
+    }
+
+    public NakadiCursor unshiftCursor(final ShiftedNakadiCursor cursor) throws InvalidCursorOperation {
+        if (cursor.getShift() < 0) {
+            return moveCursorBackwards(cursor.getTopic(), cursor.getTimeline(), cursor.getPartition(),
+                    parse(cursor.getOffset()), cursor.getShift());
```

Comment:
@v-stepanov you're right. In the `parse` method I make the assumption that the offset is a decimal zero padded number. But for `ONE` it could be hexa, right?